datapath: Enable N/S LB for overlapping pod CIDR

jibi opened the pull request:
see individual commits

aspsk committed:
bpf: fix tc_nodeport_lb4_nat_lb

The tc_nodeport_lb4_nat_lb might fail with

   --- FAIL: TestBPF/tc_nodeport_lb4_nat_lb.o (0.01s)
        bpf_test.go:172: loading collection: inlining global data: global constants must be in .rodata.config, but found reference to .bss

if SECLABEL is used inside lib/nodeport.h as an initializer for a variable.
Fix this by defining SECLABEL as an int for the test.

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nodeport: set correct src id in CT entry

when creating a new CT entry in nodeport_lb4(), set the src_sec_id field
of the new entry to the identity passed to the function rather than
defaulting to WORLD_IPV4_ID.

This change is mostly for consistency as that field is only used for
debugging purposes when the CT table is dumped.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nodeport: replace unknown src id with WORLD_IPV4_ID

when forwarding a packet in nodeport_lb4(), replace the src_sec_identity
with WORLD_IPV4_ID in case is set to unknown.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nodeport: forward real src id only when clutermesh is enabled

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi commented on bpf/lib/nodeport.h:
```diff
@@ -2709,6 +2716,10 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 
 		src_port = tunnel_gen_src_port_v4(&tuple);
 
+#if !(defined(ENABLE_CLUSTER_AWARE_ADDRESSING) && defined(ENABLE_INTER_CLUSTER_SNAT))
```
offline it was discussed to have the callers of `nodeport_lb4()` set the `src_sec_identity`, and then avoid messing with it inside the tail calls chain (@brb) but this needs to remain in `tail_nodeport_nat_egress_ipv4()` as overwriting the source identity with `WORLD_IPV4_ID` in the callers would break policy enforcement

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -880,6 +880,9 @@ tail_handle_ipv4(struct __ctx_buff *ctx, __u32 ipcache_srcid, const bool from_ho
 
+	if (src_sec_identity == 0)
+		src_sec_identity = WORLD_IPV4_ID;
+
```
Imho this change (and the corresponding one in bpf_overlay) is way too broad, and can have unexpected side-effects for existing code. Why not push this "override" logic down into `nodeport_lb4()` ?

jibi replied:
see my previous comment, @brb asked offline to have the caller of `nodeport_lb4()` set the src identity so I just followed that, but I don't mind moving it a bit deeper in the call stack

julianwiedmann replied:
hm yes, but you're changing the calling paths at a much higher level - it's not obvious that **all** subsequent users of the src_sec_identity want to have this override. What if one of them has unique handling for `identity == 0`, that now doesn't get applied any more?

I think you'll have a much easier life keeping this contained inside the nodeport-scope for now.

jibi replied:
> hm yes, but you're changing the calling paths at a much higher level

yes, on purpose in order to have a consistent value across all users rather than having it set to 0 in one code path and WORLD_IPV4_ID in another

> it's not obvious that all subsequent users of the src_sec_identity want to have this override. What if one of them has unique handling for identity == 0, that now doesn't get applied any more?

fair, perhaps I have too much faith in the tests :disappointed: 

> I think you'll have a much easier life keeping this contained inside the nodeport-scope for now.

sounds good, I'll go with that then :ok_hand:

julianwiedmann replied:
Just checking - this isn't updated in the latest push, right? The identity override is still in every calling path, not down in `nodeport_lb4()`.

jibi replied:
ah, I understood we just wanted to move it right before calling into `nodeport_lb4()` and not in that function, let me move it back down to that function then

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2928,7 +2928,7 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 		switch (ret) {
 		case CT_NEW:
 redo:
-			ct_state_new.src_sec_id = WORLD_IPV4_ID;
+			ct_state_new.src_sec_id = src_sec_identity;
 			ct_state_new.node_port = 1;
```
Minor patch ordering aspect - let's move this patch a bit further back in the series, after all the "set src identity to world if 0" changes are done?

I'm not aware that any BPF code uses the `src_sec_id` (it's definitely not returned on the CT lookup), did you check how this change would affect the agent/proxy?

jibi replied:
>I'm not aware that any BPF code uses the src_sec_id (it's definitely not returned on the CT lookup), did you check how this change would affect the agent/proxy?

yes, it just used for debugging purposes and printed when the CT table is dumped

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2661,6 +2662,8 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 	struct iphdr *ip4;
 	__s8 ext_err = 0;
 #ifdef TUNNEL_MODE
+	__u8 cluster_id __maybe_unused = (__u8)ctx_load_meta(ctx, CB_CLUSTER_ID_EGRESS);
```
The `__maybe_unused` isn't needed, correct?

jibi replied:
it's needed in case `defined(ENABLE_CLUSTER_AWARE_ADDRESSING) && defined(ENABLE_INTER_CLUSTER_SNAT)` is not true as that guards the only block where cluster_id is used

julianwiedmann replied:
Ah, otherwise we still pass it to `lookup_ip4_remote_endpoint()` but then no-one uses the result :+1: .

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2661,6 +2662,8 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 	__s8 ext_err = 0;
 #ifdef TUNNEL_MODE
+	__u8 cluster_id __maybe_unused = (__u8)ctx_load_meta(ctx, CB_CLUSTER_ID_EGRESS);
+	__u32 src_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
```
```suggestion
	__u32 src_sec_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
```

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2979,6 +2986,8 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 #endif /* DSR_ENCAP_MODE */
 		ep_tail_call(ctx, CILIUM_CALL_IPV4_NODEPORT_DSR);
 	} else {
+		ctx_store_meta(ctx, CB_SRC_LABEL, src_sec_identity);
+		ctx_store_meta(ctx, CB_CLUSTER_ID_EGRESS, cluster_id);
```
FWIW, I think it would be worth doing separate patches for the `cluster_id` and `src_sec_identity` change. They are handling different aspects - and while the `cluster_id` change is trivial, the `src_sec_identity` thing has more potential to need a revert.

julianwiedmann replied:
Sorry, I should have been clearer. I meant to extract *all* of the `cluster_id` related changes into a separate patch. Not just this single line.

julianwiedmann commented on bpf/lib/encap.h:
```diff
@@ -32,6 +34,9 @@ __encap_with_nodeid(struct __ctx_buff *ctx, __u32 src_ip, __be16 src_port,
 
+	if (seclabel > ENCAP_MAX_VNI)
+		return DROP_INVALID_IDENTITY;
+
```
To clarify - is this patch meant to address the CIDR-identity concern from https://github.com/cilium/cilium/pull/29923#discussion_r1429574175, or are you planning to add more graceful handling?

jibi replied:
for now just to make sure we don't pass down local identities. This is the only place where we return the invalid ID drop reason, so should be already a good way to detect if this happens

julianwiedmann replied:
But you could easily catch that at the specific location where `Cluster-aware-addressing` sets the forwarded identity? And thus self-document that it's not supported yet (and why).

julianwiedmann reviewed (changes requested):
Thank you! A few change requests below.

jibi committed:
bpf: nodeport: enable N/S LB for overlapping pod CIDR

This commit enables access to a K8s service from outside (N/S)
when a request first comes to a node which belongs to cluster which does
NOT run a service backend, and the overlapping pod CIDR feature is
enabled. For example:

  ┌───────────────────────┐
  │                       │
  │  Client from outside  │
  │                       │
  └──┬────────────────────┘
     │
     │ curl $NODE1:$NODEPORT
     │
  ┌──▼─────────────────────┐
  │                        │
  │  K8s Node1@Cluster1    │
  │                        │
  └──┬─────────────────────┘
     │
     │ SNAT + vxlan
     │ (overlapping pod CIDR)
     │
  ┌──▼─────────────────────┐
  │                        │
  │  K8s Node2@Cluster2    │
  │                        │
  └────────────────────────┘

To make it possible, the N/S LB entry function "nodeport_lb4()" is modified
to pass a dst cluster ID to "tail_nodeport_nat_egress_ipv4()". This
allows the latter to query a correct service endpoint in the IPcache,
and then to set an SNAT IP addr accordingly.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Anton Protopopov <aspsk@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nodeport: pass src id to tail_nodeport_nat_egress_ipv4

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nodeport: use variable for src id in nodeport_rev_dnat_ingress_ipv4

to allow alternative implementations to interact with it

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

brb reviewed (approved):
Thanks!

jibi committed:
bpf: nodeport: don't forward local CIDR identities

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann reviewed (approved):
diff looks good, ty jibi!

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -2907,6 +2914,22 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		  * only 24 bits.
+		  */
+
+		if (identity_is_cidr_range(src_sec_identity))
+			return DROP_INVALID_IDENTITY;
```
Does this make sense?

If I follow correctly, a packet coming into a NodePort from outside of the cluster could be classified by Cilium to have a CIDR identity as the source, which is completely fine, then instead of forwarding the request, we just drop it?

joestringer replied:
Hmm, I guess if we think this should never happen because of other constraints in the datapath (ie we never expect such identities to show up at this point in the processing), then as a way to do a safety check this may make sense. We may want to add `HOST_ID` to these as well, since we really shouldn't forward that identity on to another node (nor should another node blindly trust an incoming usage of the host identity).

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -2907,6 +2914,22 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+			src_sec_identity = WORLD_IPV4_ID;
+#endif
+
+		 /* Before forwarding the identity, make sure it's not a CIDR
+		  * identity, as these are __u32 values, but transporting them
+		  * via the VNI field in the VXLAN / Geneve header allows for
+		  * only 24 bits.
+		  */
```
I'm not sure that the size of the identity is the problem here. The problem is that the identity we are trying to encode into the vxlan header only has meaning locally. So even if it were possible to somehow encode the identity into the header, once the packet arrives at the next hop, that hop wouldn't understand the identity because it's not synchronized cluster-wide (or across clusters).

jibi replied:
yeah, I think I'm just quoting the implementation detail (CIDR identities aka local identities don't fit in the VNI field) rather than stating clearly that forwarding local identities doesn't make sense, but that's the base idea

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -2907,6 +2914,22 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		  * only 24 bits.
+		  */
+
+		if (identity_is_cidr_range(src_sec_identity))
```
I wonder also whether we shouldn't just be checking "in CIDR range", but actually the full upper 8 bits check (are any bits there set). Comparing this with `pkg/identity/numericidentity.go` "scopes" I also see the remote node scope, which could use a range from `2<<24` up , which would be outside the range that this function checks :thinking:
